### ISSUE TITLE
Undeploy Failed Nodes

### ISSUE THREAD

**Issue by milesstoetzner** [onecommons/unfurl#282]
I am having some troubles with the behaviour Unfurl is handling errors when a management operation fails.
I will split up my questions into multiple issues (see https://github.com/onecommons/unfurl/issues/280 and https://github.com/onecommons/unfurl/issues/281).
I dont think they are related to this issue.

If the deployment of a node fails and the application is undeployed, then Unfurl does not undeploy the failed node. 
I would expect that the corresponding delete management operation of the failed node is called. 
This gives the delete management operation the opportunity to cleanup the failed create management operation.
Of course the delete management operation must be implemented with that in mind.

What are your opinions on that?
If I am not wrong, [opera](https://github.com/xlab-si/xopera-opera) is doing that and I find it very handy.

Here is a simple `ensemble.yaml`.

```
apiVersion: unfurl/v1alpha1
kind: Ensemble
spec:
  service_template:
    node_types:
      my_host:
        derived_from: tosca:Root
        capabilities:
          host:
            type: tosca.capabilities.Compute

    topology_template:
      node_templates:
        my_host:
          type: my_host
          interfaces:
            Standard:
              operations:
                create:
                  implementation: exit 1
```

And here is the corresponding output when deploying this template.

```
(.venv) stoetzms@storm:~/unfurl-issue$ unfurl version
unfurl version 0.9.1 (24ce4686)
(.venv) stoetzms@storm:~/unfurl-issue$ unfurl deploy --approve
   INFO   UNFURL Using home project at: "/home/stoetzms/.unfurl_home/unfurl.yaml"
   INFO   UNFURL Loaded project at /home/stoetzms/unfurl-issue/unfurl.yaml
   INFO   UNFURL Using home project at: "/home/stoetzms/.unfurl_home/unfurl.yaml"
   INFO   UNFURL Loaded project at /home/stoetzms/unfurl-issue/unfurl.yaml
   INFO   UNFURL creating deploy job for /home/stoetzms/unfurl-issue/ensemble/ensemble.yaml
   INFO   UNFURL Vault password found, configuring vault ids: ['unfurl-issueVM', '.unfurl_homeF0']
   INFO   UNFURL Validating TOSCA template at /home/stoetzms/unfurl-issue/ensemble/ensemble.yaml
   INFO   UNFURL Loading ensemble "/home/stoetzms/unfurl-issue/ensemble/ensemble.yaml" in environment "None"
   INFO   TOSCA.MODEL The pre-parsed input successfully passed validation.
   INFO   UNFURL Vault password found, configuring vault ids: ['+%', '.unfurl_homeF0']
   INFO   UNFURL Validating TOSCA template at /home/stoetzms/.unfurl_home/.
   INFO   UNFURL Loading ensemble "/home/stoetzms/.unfurl_home/ensemble/ensemble.yaml" in environment "None"
   INFO   UNFURL starting deploy job for /home/stoetzms/unfurl-issue/ensemble/ensemble.yaml
   INFO   UNFURL Plan for deploy:

Node "my_host" (pending):
- operation create (reason: add)
   INFO   UNFURL.TASK Task create for my_host (reason: add): started task.
 WARNING  UNFURL.TASK Task create for my_host (reason: add): shell task run failure: "exit 1" in /home/stoetzms/unfurl-issue/ensemble
   INFO   UNFURL.TASK Task create for my_host (reason: add): shell task return code: 1, stderr:
  ERROR   UNFURL.TASK Task create for my_host (reason: add): Task failed, Resource Status: UNKNOWN State: creating
   INFO   UNFURL saving job changes to /home/stoetzms/unfurl-issue/ensemble/jobs/job2024-01-13-21-47-25-935608.yaml
   INFO   UNFURL saving changelog to /home/stoetzms/unfurl-issue/ensemble/jobs.tsv
   INFO   UNFURL
Job A41DLlPv0000 completed in 0.027s: error. 1 tasks (1 changed, 0 ok, 1 failed, 0 unknown, 0 skipped):

 ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
                   Task   Resource             Operation             Reason           Status          State               Changed
 ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
             1 (failed)   my_host              create                add              ERROR           creating            Yes
 Output: {'args': 'exit 1', 'returncode': 1, 'stdout': '', 'stderr': '', 'cmd': 'exit 1', 'timeout': None, 'error': None}

 ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Done, full log written to /home/stoetzms/unfurl-issue/ensemble/jobs/job2024-01-13-21-47-25-A41DLlPv.log
(.venv) stoetzms@storm:~/unfurl-issue$ unfurl plan
   INFO   UNFURL Using home project at: "/home/stoetzms/.unfurl_home/unfurl.yaml"
   INFO   UNFURL Loaded project at /home/stoetzms/unfurl-issue/unfurl.yaml
   INFO   UNFURL Using home project at: "/home/stoetzms/.unfurl_home/unfurl.yaml"
   INFO   UNFURL Loaded project at /home/stoetzms/unfurl-issue/unfurl.yaml
   INFO   UNFURL Vault password found, configuring vault ids: ['unfurl-issueVM', '.unfurl_homeF0']
   INFO   UNFURL Validating TOSCA template at /home/stoetzms/unfurl-issue/ensemble/ensemble.yaml
   INFO   UNFURL Loading ensemble "/home/stoetzms/unfurl-issue/ensemble/ensemble.yaml" in environment "None"
   INFO   TOSCA.MODEL The pre-parsed input successfully passed validation.
   INFO   UNFURL Vault password found, configuring vault ids: ['+%', '.unfurl_homeF0']
   INFO   UNFURL Validating TOSCA template at /home/stoetzms/.unfurl_home/.
   INFO   UNFURL Loading ensemble "/home/stoetzms/.unfurl_home/ensemble/ensemble.yaml" in environment "None"
   INFO   UNFURL creating deploy plan for /home/stoetzms/unfurl-issue/ensemble/ensemble.yaml
   INFO   UNFURL Plan for deploy:

Node "my_host" (unknown, creating):
- operation create (reason: missing)
Done, full log written to /home/stoetzms/unfurl-issue/ensemble/jobs/job2024-01-13-21-47-28-A41DLlSl.log
```

And when I try to undeploy the application: 

```
(.venv) stoetzms@storm:~/unfurl-issue$ unfurl undeploy --approve
   INFO   UNFURL Using home project at: "/home/stoetzms/.unfurl_home/unfurl.yaml"
   INFO   UNFURL Loaded project at /home/stoetzms/unfurl-issue/unfurl.yaml
   INFO   UNFURL Using home project at: "/home/stoetzms/.unfurl_home/unfurl.yaml"
   INFO   UNFURL Loaded project at /home/stoetzms/unfurl-issue/unfurl.yaml
   INFO   UNFURL creating undeploy job for /home/stoetzms/unfurl-issue/ensemble/ensemble.yaml
   INFO   UNFURL Vault password found, configuring vault ids: ['unfurl-issueVM', '.unfurl_homeF0']
   INFO   UNFURL Validating TOSCA template at /home/stoetzms/unfurl-issue/ensemble/ensemble.yaml
   INFO   UNFURL Loading ensemble "/home/stoetzms/unfurl-issue/ensemble/ensemble.yaml" in environment "None"
   INFO   TOSCA.MODEL The pre-parsed input successfully passed validation.
   INFO   UNFURL Vault password found, configuring vault ids: ['+%', '.unfurl_homeF0']
   INFO   UNFURL Validating TOSCA template at /home/stoetzms/.unfurl_home/.
   INFO   UNFURL Loading ensemble "/home/stoetzms/.unfurl_home/ensemble/ensemble.yaml" in environment "None"
   INFO   UNFURL starting undeploy job for /home/stoetzms/unfurl-issue/ensemble/ensemble.yaml
   INFO   UNFURL Plan for undeploy (workflow = undeploy):

Nothing to do.
Job A41DLoGA0000 completed: ok. No tasks ran.
Done, full log written to /home/stoetzms/unfurl-issue/ensemble/jobs/job2024-01-13-21-50-16-A41DLoGA.log
```

**Comment by aszs** [comment]
The reason it doesn't try to delete the node is similar to what's happen in #281 -- the create operation failed without setting the node into a created or error statue so Unfurl doesn't think node ever created.  Also undeploy wouldn't plan any tasks because no delete operation was defined for the node. If I update your example to set the node into an error and add a delete operation it behaves as you expect. For example:

```yaml
apiVersion: unfurl/v1alpha1
kind: Ensemble
spec:
  service_template:
    node_types:
      my_host:
        derived_from: tosca:Root
        capabilities:
          host:
            type: tosca.capabilities.Compute

    topology_template:
      node_templates:
        my_host:
          type: my_host
          interfaces:
            Standard:
              operations:
                create:
                  implementation: exit 1
                  inputs:
                    done:
                      status: error
                delete:
                  implementation: exit 0
```

`unfurl deploy -a`:
```
Job A41H5nWc0000 completed in 0.150s: error. 1 tasks (1 changed, 0 ok, 1 failed, 0 unknown, 0 skipped):

 ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
                       Task   Resource               Operation                Reason             Status            State                  Changed
 ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
                 1 (failed)   my_host                create                   add                ERROR             creating               Yes
 Output: {'args': 'exit 1', 'returncode': 1, 'stdout': '', 'stderr': '', 'cmd': 'exit 1', 'timeout': None, 'error': None}
```

`unfurl deploy -a`:
```
Job A41H5k870000 completed in 0.144s: ok. 1 tasks (1 changed, 1 ok, 0 failed, 0 unknown, 0 skipped):

 ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
                        Task   Resource               Operation               Reason                Status            State               Changed
 ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
                 1 (success)   my_host                delete                  undeploy              OK                deleted             Yes
 Output: {'args': 'exit 0', 'returncode': 0, 'stdout': '', 'stderr': '', 'cmd': 'exit 0', 'timeout': None, 'error': None}
```

**Comment by aszs** [comment]
we finally released 1.0, which includes this fix... closing this, @milesstoetzner reopen this if you are still having problems... thank you for reporting it!
